### PR TITLE
Add panos_tunnel module

### PR DESCRIPTION
**Pull request by freakinhippie**
Adding a module for creating tunnel interfaces.

### CODE REVIEW

**Comment by freakinhippie** [comment]
@DaveHewy 

This PR has been added to in progress for 2.0.2. I think that means it will be merged soon, but I'm not certain.

**Comment by DaveHewy** [comment]
@freakinhippie ok great. Does this PR solve creation of 'Network' templates/things in a Panorama setup? I got slightly worried by the following line in the PR;

``module.fail_json(msg="tunnel interfaces don't exist on Panorama")``

**Comment by freakinhippie** [comment]
@DaveHewy I am not sure. I don't have a Panorama device to test against. The module is based heavily on the panos_interface module, but the code in that module says that interfaces are not supported on Panorama. If the panos_interface module supports Panorama, then I would think this one will likely support it as well.

**Comment by DaveHewy** [comment]
@freakinhippie thanks for the reply! I'll kick the tyres against a Panorama device tomorrow using this PR & report back.

**Comment by freakinhippie** [comment]
@DaveHewy Thanks Dave. Any feedback is welcome. If I can make it support Panorama, I'm happy to do it.

**Comment by freakinhippie** [comment]
@DaveHewy Hi Dave, I managed to get read-only access to a Panorama system and can confirm that Panorama does in fact support tunnel interfaces. Unfortunately, I can also confirm that this module does not support Panorama templates.

I'm trying to get my hands on a Panorama VM that I can use for testing so that I can extend the support but at the moment I don't have access to one.

**Comment by DaveHewy** [comment]
hi @freakinhippie 
I have tested this against Panorama and now have a working version of the tunnel interface. Albeit a diverged from upstream version.

The key seems to be in the below lines. Which is ultimately responsible for switching the xpath context (somewhat deeper in panorama).

```
    # Open the connection to the PANOS device.
    device = None
    try:
        device = PanDevice.create_from_device(*auth)
    except PanDeviceError:
        e = get_exception()
        module.fail_json(msg=e.message)

    # Set the attachment point for the Zone object
    parent = None
    if isinstance(device, Firewall):
        parent = device
    elif isinstance(device, Panorama):
        if template is not None:
            template_list = Template.refreshall(device)
            parent = get_template(template, template_list)
            if parent is None:
                module.fail_json(msg='Template not found: {0}'.format(template))
        else:
            module.fail_json(msg='A template parameter is required when device type is Panorama')
    if vsys is not None:
        v = Vsys(vsys)
        parent.add(v)
        parent = v
```

You would then continue to use `parent` instead of `device`
Its worth looking into the xpaths used for both to see the differences

**Comment by stealthllama** [comment]
You can spin up a Panorama instance on AWS.  https://aws.amazon.com/marketplace/pp/B07BNKJM2H/ref=portal_asin_url

**Comment by stealthllama** [comment]
@freakinhippie:  I'm actually working on VLAN support right now and that will include VLAN interfaces (VRI).  I exchanged with @ivanbojer and he'd prefer to see VLAN, tunnel, and loopback interfaces included in the existing panos_interface module.  I'm in the process of Panorama-ifying that now.  I can take what you've done for panos_tunnel and incorporate that as well.

**Comment by pixelicous** [comment]
please merge this :)

**Comment by freakinhippie** [comment]
@stealthllama, that sounds great! Thanks

**Comment by geraint-jones-nz** [comment]
I would suggest merging this until a unified `panos_interfaces` module is done.

**Comment by shinmog** [comment]
I'll chat with @ivanbojer about this

File: library/panos_tunnel.py [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+
+    if commit:
+        try:
+            con.commit(sync=True)
+        except errors.PanDeviceError:
+            e = get_exception()
+            module.fail_json(msg='Performed {0} but commit failed: {1}'.format(state, e.message))
+
+    # Done!
+    module.exit_json(changed=True, msg='okey dokey')
```
Comment (shinmog) [inline library/panos_tunnel.py]: Heh, just say "Done" here :)

File: library/panos_tunnel.py [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+def main():
+    argument_spec = dict(
+        ip_address=dict(required=True),
+        password=dict(no_log=True),
+        username=dict(default='admin'),
+        api_key=dict(no_log=True),
+        state=dict(default='present', choices=['present', 'absent']),
+        if_name=dict(required=True),
+        ip=dict(type='list'),
+        ipv6_enabled=dict(),
```
Comment (shinmog) [inline library/panos_tunnel.py]: `type='bool'`

File: library/panos_tunnel.py [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+        management_profile=dict(),
+        mtu=dict(),
```
Comment (shinmog) [inline library/panos_tunnel.py]: `type='int'`

Comment (freakinhippie) [reply library/panos_tunnel.py]: @shinmog I've cleaned this up as requested. Thanks!

**Pull request merged**